### ISSUE TITLE
Proper support for s390x

### ISSUE THREAD

**Issue by samoht** [mirage/irmin#1693]
The tests seem to segfault on s390x (https://github.com/ocaml/opam-repository/pull/20033#issuecomment-969439995):

```
#=== ERROR while compiling irmin.2.9.0 ========================================#
# context              2.1.1 | linux/s390x | ocaml-base-compiler.4.13.1 | pinned(https://github.com/mirage/irmin/releases/download/2.9.0/irmin-2.9.0.tbz)
# path                 ~/.opam/4.13/.opam-switch/build/irmin.2.9.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune runtest -p irmin -j 3
# exit-code            1
# env-file             ~/.opam/log/irmin-1833-af4727.env
# output-file          ~/.opam/log/irmin-1833-af4727.out
### output ###
#         test alias test/irmin/runtest (got signal SEGV)
# (cd _build/default/test/irmin && ./test.exe)
# Testing `irmin'.
# This run has ID `XSEPBKU1'.
# 
#   [OK]          tree          0   bindings.
#   [OK]          tree          1   paginated bindings.
#   [OK]          tree          2   diff.
#   [OK]          tree          3   add.
#   [OK]          tree          4   remove.
#   [OK]          tree          5   update.

-         test alias test/irmin/runtest (got signal SEGV)
- (cd _build/default/test/irmin && ./test.exe)
- Testing `irmin'.
- This run has ID `XSEPBKU1'.
- 
-   [OK]          tree          0   bindings.
-   [OK]          tree          1   paginated bindings.
-   [OK]          tree          2   diff.
-   [OK]          tree          3   add.
-   [OK]          tree          4   remove.
-   [OK]          tree          5   update.
```

We should investigate and fix this for 2.10

**Comment by samoht** [comment]
The issue comes from the OCaml compiler: https://github.com/ocaml/ocaml/issues/10857

Tail-call recursive functions with too many arguments are not properly optimised, and the limit for s390x is very low. The good news is that it is fixed in ocaml's master branch (which will become 4.14 at one point). It is a bit unclear how we should fix this for 4.13, as there are probably other places where the compilation of tail-rec functions is incorrect.

**Comment by CraigFe** [comment]
I think we can probably close this, as https://github.com/mirage/irmin/pull/1694 appears to fix the immediate problem on s390x (although it's likely there are other issues beneath the surface). OCaml 4.14 is the proper solution, and this is not far away.
